pkg/cloudevents/types/value.go

```diff
@@ -10,201 +12,212 @@ import (
-		return fmt.Sprintf("<unknown:%T>", v.v)
+	rx := reflect.ValueOf(v)
+	if rx.Kind() == reflect.Ptr && !rx.IsNil() {
+		// Allow pointers-to convertible types
+		return Validate(rx.Elem().Interface())
 	}
+	return nil, fmt.Errorf("invalid CloudEvents value: %#v", v)
 }
 
-// ToBool returns a bool value, decoding from string if necessary.
-func (v Value) ToBool() (bool, error) {
-	switch x := v.v.(type) {
+// ToBool accepts a bool value or canonical "true"/"false" string.
+func ToBool(v interface{}) (bool, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case bool:
-		return x, nil
+		return v, nil
 	case string:
-		return strconv.ParseBool(x)
+		return ParseBool(v)
 	default:
-		return false, typeErr("Bool", x)
+		return false, convertErr("Bool", v, err)
 	}
 }
 
-// ToInteger returns an int32 value, decoding from string if necessary.
-func (v Value) ToInteger() (int32, error) {
-	switch x := v.v.(type) {
+// ToInteger accepts any numeric value in int32 range, or canonical string.
+func ToInteger(v interface{}) (int32, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case int32:
-		return x, nil
+		return v, nil
 	case string:
-		// Accept floating-point but truncate to int32 as per CE spec.
-		f, err := strconv.ParseFloat(x, 64)
-		if f > math.MaxInt32 || f < math.MinInt32 {
-			return 0, rangeErr(x)
-		}
-		return int32(f), err
+		return ParseInteger(v)
 	default:
-		return 0, typeErr("Integer", x)
+		return 0, convertErr("Integer", v, err)
 	}
 }
 
-// ToString returns a string value. It does NOT encode non-string, use String() for that.
-func (v Value) ToString() (string, error) {
-	switch x := v.v.(type) {
+// ToString returns a string value unaltered.
+//
+// This function does not perform canonical string encoding, use one of the
+// Format functions for that.
+func ToString(v interface{}) (string, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case string:
-		return x, nil
+		return v, nil
 	default:
-		return "", typeErr("String", x)
+		return "", convertErr("String", v, err)
 	}
 }
 
-// ToBinary returns a []byte value, decoding from string if necessary.
-func (v Value) ToBinary() ([]byte, error) {
-	switch x := v.v.(type) {
+// ToBinary returns a []byte value, decoding from base64 string if necessary.
+func ToBinary(v interface{}) ([]byte, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case []byte:
-		return x, nil
+		return v, nil
 	case string:
-		return base64.StdEncoding.DecodeString(x)
+		return base64.StdEncoding.DecodeString(v)
 	default:
-		return nil, typeErr("Binary", x)
+		return nil, convertErr("Binary", v, err)
 	}
 }
 
-// ToURIRef returns a url.URL value, decoding from string if necessary.
-func (v Value) ToURIRef() (url.URL, error) {
-	switch x := v.v.(type) {
-	case url.URL:
-		return x, nil
+// ToURL returns a *url.URL value, parsing from string if necessary.
+func ToURL(v interface{}) (*url.URL, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
+	case *url.URL:
```

Comment:
might be nice to also handle `url.URL`

Reply:
It is handled - Validate() does all legal conversions, so the switch only needs to consider *url.URL and string. Aside from actual validation, that's the main role of Validate() - to simplify type switches so they only have to consider the preferred native types.